### ISSUE TITLE
reducing or matching on terms in Ltac forgets type information

### ISSUE THREAD

**Issue by andres-erbsen** [coq/coq#7425]
#### Version

~~~~~
The Coq Proof Assistant, version 8.8.0 (May 2018)
compiled on May 3 2018 7:11:49 with OCaml 4.06.0
~~~~~

#### Operating system

Arch Linux, up to date as of today.

#### Description of the problem

I believe it is known that Ltac matches strip type casts -- I have in fact used this to speed up recursion under binders using typeclasses in 8.4.
It is probably also known that calling cbv from Ltac strips type casts.
Nevertheless, the last example below failing is very counter-intuitive to me; at the very least the successor of Ltac should not exhibit it.

~~~~~coq
Require Coq.Lists.List.

Goal forall (T : Type), False.
  intros.

  (* succeeds as expected *)
  let l := constr:(@cons Type nat (cons T nil)) in
  let l' := match l with @cons ?T ?x ?tl => constr:(@cons T x tl) end in
  unify l l'.

  (* fails unexpectedly, but eh, it's type inference... *)
  (
    let l := constr:(@cons Type nat (cons T nil)) in
    let l' := match l with cons ?x ?tl => constr:(cons x tl) end in
    unify l l'
  ) || idtac.

  (* fails unexpectedly, and I see no excuse for this one *)
  let l := constr:(@cons Type nat (cons T nil)) in
  let x := eval cbv in (List.hd (unit : Type) l) in
  let tl := eval cbv in (List.tl l) in
  let l' := constr:(cons x tl) in
  unify l l'.
Abort.

Set Universe Polymorphism.
(* the same repeats ... *)
Goal forall (T : Type), False.
  intros.

  (* succeeds as expected *)
  let l := constr:(@cons Type nat (cons T nil)) in
  let l' := match l with @cons ?T ?x ?tl => constr:(@cons T x tl) end in
  unify l l'.

  (* fails unexpectedly, but eh, it's type inference... *)
  (
    let l := constr:(@cons Type nat (cons T nil)) in
    let l' := match l with cons ?x ?tl => constr:(cons x tl) end in
    unify l l'
  ) || idtac.

  (* fails unexpectedly, and I see no excuse for this one *)
  let l := constr:(@cons Type nat (cons T nil)) in
  let x := eval cbv in (List.hd (unit : Type) l) in
  let tl := eval cbv in (List.tl l) in
  let l' := constr:(cons x tl) in
  unify l l'.
Abort.
~~~~~

**Comment by andres-erbsen** [comment]
Perhaps the issue is not that ltac forgets types, but that they change at all? I would expect the type of `hd` from a `list Type` to be convertible usable as `Type`, and currently it is not. The mechanisms behind the current situation in Ltac may be intentional for once, but I think the example here is actually problematic in practice. In particular, the workaround whose necessity provoked me to open this issue looks rather silly on its own:

~~~~~
Ltac list_union l1 l2 :=
  lazymatch l1 with
  | nil => l2
  | @cons ?T ?x ?l1' =>
    let x := (* preserve types after matching, COQBUG(7425) *)
        match type of x with
        | T => x
        | _ => constr:((x : T))
        end in
    let l2 := add_to_set x l2 in
    list_union l1' l2
  end.
~~~~~

However, I think another sufficient criterion for closing this issue would be an argument for why something like this will never happen after template polymorphism is properly rooted out.

**Comment by SkySkimmer** [comment]
This isn't a template polymorphism problem, it's losing the information that a type was used at a higher universe through cumulativity.

**Comment by andres-erbsen** [comment]
Okay, I talked to @JasonGross about this and we arrived at another interpretation of the situation. Basically, I should not expect to write tactics that operate over polymorphic lists but do not take the type parameter of these lists as an argument. Concretely, `list_union` should be passing `T` to `add_to_set`, which would use pass it to `constr:(@cons ...)`, instead of expecting the `nat` to carry the type or for it to be re-inferred. And for the cbv it is possible to have the following hack:

~~~~~
Ltac cbv_at_same_type x :=
  let x := eval cbv in x in
  let T := match type of x with ?T => T end in
  constr:(x : T)
~~~~~
